tsdb/engine/tsm1/engine.go

```diff
@@ -228,7 +258,8 @@ func (e *DevEngine) Begin(writable bool) (tsdb.Tx, error) {
 
 func (e *DevEngine) WriteTo(w io.Writer) (n int64, err error) { panic("not implemented") }
 
-func (e *DevEngine) writeSnapshot() error {
+// WriteSnapshot will snapshot the cache and write a new TSM file with its contents, releasing the snapshot when done.
+func (e *DevEngine) WriteSnapshot(async bool) error {
```

Comment:
@pauldix That is very different code then is what is here now.  I would like to see data that the current code exhibits the same issue.  If it is CPU bound, it's very likely we can tune it to reduce CPU.  The most CPU intensive part of writing a snapshot _should_ be the encoding which has been tuned quite a bit and should support a rate much faster than we're snapshotting.  If it's not the encoding, then we need to examine that first.  Throwing more goroutines at the problem is not right solution based on what I have seen.  In fact, it creates new problems with contention, potential race conditions as well as makes the code harder to reason about.

In regards to panics, the code should absolutely not be panicing in these cases.  We need to design a solution that handles errors appropriately.